ocaml/testsuite/tests/typing-layouts/basics.ml

```diff
@@ -2331,3 +2331,39 @@ Error: This expression has type t_float64
        But the layout of t_float64 must be a sublayout of value, because
          of the definition of t40 at line 1, characters 0-16.
 |}]
+
+(**********************************************************************)
+(* Test 41: constraints in manifests in mutually recursive typedecls. *)
+
+(* This example must be rejected. *)
+type t1 = string t2 as (_ : immediate)
+and 'a t2 = 'a
+
+[%%expect{|
+Line 2, characters 0-14:
+2 | and 'a t2 = 'a
+    ^^^^^^^^^^^^^^
+Error:
+       The layout of 'a t2 is value, because
+         it instantiates an unannotated type parameter of t2, defaulted to layout value.
+       But the layout of 'a t2 must be a sublayout of immediate, because
+         of the annotation on the wildcard _ at line 1, characters 28-37.
+|}]
+
+(* This example is unfortunately rejected as a consequence of the fix for the
+   above in typedecl. If we ever change that so that the below starts working,
+   make sure [t1]'s parameter is immediate! Previously this was allowed and t1's
+   parameter was just value (a bug). *)
+type 'a t1 = 'a t2 as (_ : immediate)
+and 'a t2 = 'a
+
+[%%expect{|
+Line 2, characters 0-14:
+2 | and 'a t2 = 'a
+    ^^^^^^^^^^^^^^
+Error:
+       The layout of 'a t2 is value, because
```

Comment:
This message would be better to admit that the compiler is not clever enough to accept the definition. Don't know how hard that would be to arrange.

Reply:
Arranged - see below.